zilliqa/src/consensus.rs

```diff
@@ -2718,150 +2738,178 @@ impl Consensus {
             trace!("applying {} transactions to state", transactions.len());
         }
 
-        let mut verified_txns = Vec::new();
-
-        // We re-inject any missing Intershard transactions (or really, any missing
-        // transactions) from our mempool. If any txs are unavailable in both the
-        // message or locally, the proposal cannot be applied
-        for (idx, tx_hash) in block.transactions.iter().enumerate() {
-            // Prefer to insert verified txn from pool. This is faster.
-            let txn = match self.transaction_pool.pop_transaction(*tx_hash) {
-                Some(txn) => txn,
-                _ => match transactions
-                    .get(idx)
-                    .map(|sig_tx| sig_tx.clone().verify())
-                    .transpose()?
-                {
-                    // Otherwise, recover txn from proposal. This is slower.
-                    Some(txn) if txn.hash == *tx_hash => txn,
-                    _ => {
-                        warn!("Proposal {} at view {} referenced a transaction {} that was neither included in the broadcast nor found locally - cannot apply block", block.hash(), block.view(), tx_hash);
-                        return Ok(());
-                    }
-                },
-            };
-            verified_txns.push(txn);
-        }
+        let mut block_receipts: Vec<(TransactionReceipt, usize)> = Vec::new();
 
-        let mut block_receipts = Vec::new();
-        let mut cumulative_gas_used = EvmGas(0);
-        let mut receipts_trie = eth_trie::EthTrie::new(Arc::new(MemoryDB::new(true)));
-        let mut transactions_trie = eth_trie::EthTrie::new(Arc::new(MemoryDB::new(true)));
+        if from.is_some_and(|peer_id| peer_id == self.peer_id()) {
```

Comment:
Is it worth checking here that we have the correct number of transaction receipts in memory here in case the node has restarted in-between broadcasting the proposal and calling this code? 

Similarly, we may begin building this data without using it, say if we think were leader but there is a fork and the other branch wins. In this case should we nullify our in-memory receipts and add the transactions back to the mempool?

Reply:
Good idea. I'll implement a flush to the receipts cache.